### DISCUSSION TITLE
how custom discreate colormap work with expression

### DISCUSSION THREAD

**Discussion by satas12825** [cogeotiff/rio-tiler#721] [Q&A]
I'm attempting to create a custom colormap for my NDVI raster, but I think I might have misunderstood something, as the results aren't quite what I expected. Could you please help me figure out what might be going wrong?

first i try 
ndvi_cmap = {
    -1: (0, 0, 255, 255),   # Blue for non-vegetation
    -0.5: (0, 255, 255, 255),  # Cyan for barren land
    0: (255, 255, 0, 255),  # Yellow for sparse vegetation
    0.2: (200, 255, 0, 255),  # Light green for low vegetation
    0.4: (100, 255, 0, 255),  # Green for moderate vegetation
    0.6: (0, 255, 0, 255),  # Dark green for dense vegetation
    1: (0, 128, 0, 255)     # Very dark green for very dense vegetation
}

then try  to rescale value to 0-255 
            ndvi_cmap = {
                0: (0, 0, 255, 255),   # Blue for non-vegetation (e.g., water)
                63: (0, 255, 255, 255),  # Cyan for barren land or built-up areas
                127: (255, 255, 0, 255),  # Yellow for sparse vegetation
                153: (200, 255, 0, 255),  # Light green for low vegetation
                178: (100, 255, 0, 255),  # Green for moderate vegetation
                204: (0, 255, 0, 255),  # Dark green for dense vegetation
                255: (0, 128, 0, 255)     # Very dark green for very dense vegetation
            }

and then use colormap like this
            expression="(b3-b1)/(b1+b3)"
            colormap = cmap
            # Register the custom NDVI colormap
            colormap = colormap.register({"ndvi_cmap": ndvi_cmap})

            cm = colormap.get('ndvi_cmap')
            with Reader(url) as cog:
                img = cog.tile(x, y, z,tilesize=512,expression=expression)
                # img.rescale(in_range=((-1,1,))
                img.rescale(in_range=((-1, 1),),)
            content = img.render(colormap=cm,img_format="PNG")
            return content

**Comment by vincentsarago** [comment]
>  as the results aren't quite what I expected

what is not expected in your results? 

I think what you want to use is not discrete colormap but interval colormap https://cogeotiff.github.io/rio-tiler/colormap/#intervals-colormaps 
```
[
    ((-1, -0.5), (0, 0, 255, 255)), # Blue for non-vegetation
    ((-0.5, 0), (0, 255, 255, 255)), # Cyan for barren land
    ((0, 0.2), (255, 255, 0, 255)), # Yellow for sparse vegetation
    ((0.2, 0.4), (200, 255, 0, 255)), # Light green for low vegetation
    ((0.4, 0.6), (100, 255, 0, 255)), # Green for moderate vegetation
    ((0.6, 1), (0, 255, 0, 255)), # Dark green for dense vegetation
    ((1, 2), (0, 128, 0, 255)) # Very dark green for very dense vegetation
]
```

or maybe make a `linear` colormap 

```python
import bisect

cm = {
    0: (0, 0, 255, 255), # Blue for non-vegetation (e.g., water)
    63: (0, 255, 255, 255), # Cyan for barren land or built-up areas
    127: (255, 255, 0, 255), # Yellow for sparse vegetation
    153: (200, 255, 0, 255), # Light green for low vegetation
    178: (100, 255, 0, 255), # Green for moderate vegetation
    204: (0, 255, 0, 255), # Dark green for dense vegetation
    255: (0, 128, 0, 255) # Very dark green for very dense vegetation
}

keys = list(cm)
ndvi_dict = {
    ii: cm[keys[bisect.bisect(keys, ii) - 1]]
    for ii in range(0, 255, 1)
} 
ndvi_dict[255] = (0, 128, 0, 255)

cmap = cmap.register({"ndvi": ndvi_dict})
```

**Reply by satas12825** [reply]
[
    ((-1, -0.5), (0, 0, 255, 255)), # Blue for non-vegetation
    ((-0.5, 0), (0, 255, 255, 255)), # Cyan for barren land
    ((0, 0.2), (255, 255, 0, 255)), # Yellow for sparse vegetation
    ((0.2, 0.4), (200, 255, 0, 255)), # Light green for low vegetation
    ((0.4, 0.6), (100, 255, 0, 255)), # Green for moderate vegetation
    ((0.6, 1), (0, 255, 0, 255)), # Dark green for dense vegetation
    ((1, 2), (0, 128, 0, 255)) # Very dark green for very dense vegetation
]

I already tried this, but the entire image turned out transparent.

**Reply by vincentsarago** [reply]
for this colormap, make sure you don't `rescale` your image.

**Comment by satas12825** [comment]
Thank you for your response and for the great work on this project. I'm a bit confused about creating a custom colormap, and I would appreciate your guidance.

For example, if I have a multiband image, how should I specify the value range to set the colors? Specifically:

Should I set the range using values after rescaling, such as with an expression like expression="(b3-b1)/(b1+b3)", and then assign the color red to the range (0,1)?
Or should I set the range using values after rescaling but convert it to a 0-255 range?
Thank you for your assistance.
